Add skeleton and stubbing

byronlin-bq opened the pull request.

matthewh-BQ commented on cryptohasher.go:
```diff
@@ -0,0 +1,86 @@
+	hashAlgorithm hash.Hash
+}
+
+func NewCryptoHasher(algorithm string) IonHasher {
```
Are we able to enforce algorithm type by enforcing a known enum or are we going forward with passing a string value?

If it's going to be a string, I would suggest adding a error type: `func NewCryptoHasher(algorithm string) (IonHasher, error)`.

byronlin-bq replied:
Updated algorithm type to enum.

matthewh-BQ commented on cryptohasherprovider.go:
```diff
@@ -0,0 +1,29 @@
+	algorithm string
+}
+
+func NewCryptoHasherProvider(algorithm string) *cryptoHasherProvider{
```
nit: Add a space between *cryptoHasherProvider{

byronlin-bq replied:
Updated

matthewh-BQ commented on cryptohasherprovider.go:
```diff
@@ -0,0 +1,29 @@
+	return &cryptoHasherProvider{algorithm: algorithm}
+}
+
+func (chp *cryptoHasherProvider) newHasher() IonHasher{
```
nit: Similar to above

byronlin-bq replied:
Updated

Armanbqt commented on cryptohasherprovider.go:
```diff
@@ -0,0 +1,29 @@
+
+type cryptoHasherProvider struct {
+	IonHashProvider
+
```
Not sure what is the proper styling here (the empty line)

byronlin-bq replied:
I ran code formatter for go (gofmt) and it didnt complain about this. Also I see the same formatting in ion-go: https://github.com/fernomac/ion-go/blob/master/binaryreader.go#L9-L10

matthewh-BQ commented on cryptohasher.go:
```diff
@@ -0,0 +1,86 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
```
nit: newline after license. Matches style of existing Ion Go repo.

byronlin-bq replied:
Added newline for every file

matthewh-BQ commented on cryptohasherprovider.go:
```diff
@@ -0,0 +1,29 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
```
nit: Similar to first comment, please check all files and add newline.

byronlin-bq replied:
resolved

Armanbqt commented on structserializer.go:
```diff
@@ -0,0 +1,44 @@
+
+type structSerializer struct {
+	serializer
+
```
So ``serializer`` is referring to ``serializer.go`` in this package and ``Seializer`` on line 20 is the interface in that class?

byronlin-bq replied:
serializer is the abstract "class" that contains implemented functionality shared between scalarSerializer and structSerializer. 

serializer, scalarSerializer, structSerializer all implement the Serializer interface

matthewh-BQ commented on err.go:
```diff
@@ -0,0 +1,24 @@
+/*
```
We might need to read more into this, but I believe errors are handled differently in go. I imagine we'd have a generic errors.go.

Looking at MongoDB and BigQuery, I see a similar pattern and maybe this might be more idiomatic to go. I understand this is WIP, so I don't expect something fully fleshed out.

https://github.com/googleapis/google-cloud-go/blob/master/bigquery/error.go
https://github.com/mongodb/mongo-go-driver/blob/master/mongo/errors.go

justing-bq replied:
What we have here is still consistent with idiomatic Go. 
InvalidOperationError and SyntaxError implement Go's error interface. It's easy to miss this because there is no explicit syntax specifying that like in other languages. Over time we can add more error types to this file as needed.

matthewh-BQ commented on hashreader.go:
```diff
@@ -0,0 +1,144 @@
+)
+
+type HashReader interface {
+	// Extend interface of Ion reader.
```
nit: I believe this should be embed instead of extend.

byronlin-bq replied:
Updated to embed

pbcornell commented on hashreader.go:
```diff
@@ -0,0 +1,164 @@
+	return hashReader.Annotations()
+}
+
+func (hashReader *hashReader) StepIn() error {
```
needs an Ion type parameter....

byronlin-bq replied:
This function signature is based on ion.Reader stepIn function: https://github.com/fernomac/ion-go/blob/master/reader.go#L97

pbcornell replied:
oops, sorry, my mistake

pbcornell commented on hashreader.go:
```diff
@@ -0,0 +1,164 @@
+	panic("implement me")
+}
+
+// The following implements HashValue interface.
+
+func (hashReader hashReader) GetFieldName() string {
+	panic("implement me")
+}
+
+func (hashReader hashReader) GetAnnotations() []string {
+	panic("implement me")
+}
+
+func (hashReader *hashReader) Value() interface{} {
+	panic("implement me")
+}
+
+func (hashReader *hashReader) IsInStruct() bool {
+	panic("implement me")
+}
```
these should not be public

byronlin-bq replied:
updated to private

pbcornell commented on hashvalue.go:
```diff
@@ -0,0 +1,36 @@
+import ion "ion-go"
+
+//TODO add docstrings
+type HashValue interface {
```
this should not be part of the public API

byronlin-bq replied:
updated to private

pbcornell commented on hashwriter.go:
```diff
@@ -0,0 +1,189 @@
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) BeginList() error {
+	hashWriter.BeginList()
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) EndList() error {
+	hashWriter.EndList()
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) BeginSexp() error {
+	hashWriter.BeginSexp()
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) EndSexp() error {
+	hashWriter.EndSexp()
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) BeginStruct() error {
+	hashWriter.BeginStruct()
+	panic("implement me")
+}
+
+func (hashWriter *hashWriter) EndStruct() error {
+	hashWriter.EndStruct()
+	panic("implement me")
+}
```
replace with the more generic `stepIn(IonType)` and `stepOut()` pattern

byronlin-bq replied:
This is currently based on ion.Writer interface:

https://github.com/fernomac/ion-go/blob/master/writer.go#L96-L112

pbcornell commented on hashwriter.go:
```diff
@@ -0,0 +1,189 @@
+	panic("implement me")
+}
+
+// The following implements HashValue interface.
+
+func (hashWriter hashWriter) GetFieldName() string {
+	panic("implement me")
+}
+
+func (hashWriter hashWriter) GetAnnotations() []string {
+	panic("implement me")
+}
+
+func (hashWriter hashWriter) IsNull() bool {
+	panic("implement me")
+}
+
+func (hashWriter hashWriter) Type() ion.Type {
+	panic("implement me")
+}
+
+func (hashWriter hashWriter) Value() interface{} {
+	panic("implement me")
+}
+
+func (hashWriter hashWriter) IsInStruct() bool {
+	panic("implement me")
+}
```
not part of the public API

byronlin-bq replied:
updated to private

pbcornell commented on ionhashprovider.go:
```diff
@@ -0,0 +1,21 @@
+
+package ionhash
+
+type IonHashProvider interface {
```
rename to "IonHasherProvider"

pbcornell commented on serializer.go:
```diff
@@ -0,0 +1,119 @@
+import ion "ion-go"
+
+//TODO add docstrings
+type Serializer interface {
```
this should not be part of the public API

byronlin-bq replied:
Updated to serializer

pbcornell commented on serializer.go:
```diff
@@ -0,0 +1,119 @@
+
+	scalar(ionValue interface{})
+
+	stepIn(ionValue interface{})
```
`stepIn(ionType)` ?

byronlin-bq replied:
Is there a reason to pass in IonType instead of IonValue?

pbcornell replied:
nope, my mistake, sorry

byronlin-bq commented on cryptohasher.go:
```diff
@@ -0,0 +1,86 @@
+func NewCryptoHasher(algorithm string) IonHasher {
+	var hashAlgorithm hash.Hash
+
+	switch algorithm {
```
Added default to return error instead of panic().

Panic() will terminate the program which I don't think would be appropriate for this instance, we should return an error and allow the user to catch the error

byronlin-bq commented on stack.go:
```diff
@@ -0,0 +1,39 @@
+}
+
+// Push element into stack
+func (s *stack) push(element interface{}) {
```
Updated

Armanbqt commented on hashwriter.go:
```diff
@@ -0,0 +1,189 @@
+package ionhash
+
+import (
+	"ion-go"
```
You might want to reorganize the imports:
https://github.com/golang/go/wiki/CodeReviewComments#imports

byronlin-bq replied:
Addressed

Armanbqt commented on hashreader.go:
```diff
@@ -0,0 +1,164 @@
+package ionhash
+
+import (
+	"ion-go"
```
You might want to reorganize the imports:
https://github.com/golang/go/wiki/CodeReviewComments#imports

byronlin-bq replied:
Addressed

byronlin-bq commented:
> Been looking through Go conventions and looking at how Google and MongoDB does their file structuring. Seems everything is in lower case for filenames and they use a _. e.g., crypto_hasher.go.

Updated file names

byronlin-bq commented:
> 1- For naming conventions, I see:
> 
> > If the functions with names that start with an uppercase letter will be exported to other packages. If the function name starts with a lowercase letter, it won't be exported to other packages, but you can call this function within the same package.

Reviewed all function names and updated exportable functions accordingly.


> 
> There are some `new` and `New` let's make sure they are all correct.
> 

Updated.

> 2- In some structs, you've used anonymous fields, I personally think it makes it easier to read the code if we have name for all the fields.

Some structs have anonymous fields for the purpose of embedding, 

eg: 
`type structSerializer struct {
	baseSerializer
...
}
`
structSerializer is embedding baseSerializer so it 'inherits' baseSerializer's functions

byronlin-bq commented on hasher.go:
```diff
@@ -0,0 +1,51 @@
+	panic("implement me")
+}
+
+func (h *hasher) stepIn() {
```
Did you mean Ion value parameter? I see in Dotnet hash, it is passing an Ion value: https://github.com/amzn/ion-hash-dotnet/blob/master/Amazon.IonHashDotnet/Hasher.cs#L41

pbcornell replied:
yes, sorry:  value

byronlin-bq commented:
> Generally looks good; a couple overall thoughts:
> 
> * remove "2020" from copyright headers

Removed.

> * now's a good time to agree on some consistent approach to code formatting (e.g., spacing in function signatures varies a bit in this PR)

Applied code formatting (gofmt)

> * if we're following [amzn/ion-go#12](https://github.com/amzn/ion-go/issues/12), a number of HashReader's / HashWriter's functions are part of an Event, which should allow the HashReader/HashWriter interfaces to be simpler

Currently we are basing on the fernomac APIs, once the PR ^ is approved, we will update accordingly.

pbcornell commented on stack.go:
```diff
@@ -0,0 +1,40 @@
+type stack []interface{}
+
+// Check if stack is empty
+func (s *stack) IsEmpty() bool {
+	return len(*s) == 0
+}
+
+// Push element into stack
+func (s *stack) Push(element interface{}) {
+	*s = append(*s, element)
+}
+
+// Remove and return top element of stack. Return false if stack is empty.
+func (s *stack) Pop() (interface{}, error) {
```
not sure whether it matters (since stack is internal), but perhaps these functions should be lower-case as well?

byronlin-bq replied:
Updated to lower case

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,595 @@
+	"bytes"
+	"github.com/amzn/ion-go/ion"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
The github imports should be separated from the other imports.

byronlin13 replied:
Done

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,595 @@
+	provider        *testIonHasherProvider
+}
+
+func writeToWriter(t *testing.T, reader ion.Reader, textWriter ion.Writer, binaryWriter ion.Writer) {
```
I suggest moving this function to test_utilities.go where we have a similar function `writeFromReaderToWriter()` defined.

Likewise with any other helper functions that might be useful for tests outside of ion_hash_test.go.

byronlin13 replied:
Addressed

justing-bq replied:
Your latest commit created a new file `test_utilities.go` rather than adding to the already existing `testing_utilities.go`

justing-bq replied:
Pardon me. My original comment should have said `testing_utilities.go`

byronlin13 replied:
I realized I didnt merge upstream, updated and removed test_utilities.go

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,595 @@
+			require.NoError(t, err)
+			err = binaryWriter.FieldName(reader.FieldName())
+			require.NoError(t, err)
+		}
```
Looks like this FieldName logic applies to Null and all Scalar types.

How about we put this logic in a `if ionType == ion.NullType || ion.IsScalar(ionType)` block before the `switch ionType` block rather than repeating it for each of those types.

byronlin13 replied:
FieldName applies to all types if they are in a struct already. I am trying to find a way so it doesn't get repeated

byronlin13 replied:
Updated

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,595 @@
+			switch intSize {
+			case ion.Int32:
+
+				intValue, err := reader.IntValue()
```
If I recall correctly, when intSize is Int32, we can still call `reader.Int64Value()` to get the 64 bit version of the int value and pass that directly to `WriteInt()` instead of calling `reader.IntValue()` to get the 32 bit value and then converting it to 64 bit when passing to `WriteInt()`.

Because of this, we can combine Int32 and Int64 into one case.

Take a look at the `switch IntSize` in `writeFromReaderToWriter()` in `testing_utilities.go` to see another example of this.

byronlin13 replied:
Oh wow, I didnt realize writeFromReaderToWriter() existed, thanks Ill take a look

byronlin13 replied:
Updated

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,595 @@
+						annotations := reader.Annotations()
+
+						if len(annotations) > 0 {
+							if annotations[0] == "update" {
```
How about a `switch annotations[0]` instead?

byronlin13 replied:
Addressed

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,536 @@
+					}
+					err = reader.StepOut()
+					assert.NoError(t, err, "Something went wrong executing reader.StepOut()")
+				}
```
The only difference between the `fieldName == "identity"` case and `fieldName == "md5"` case is the "final_digest" logic. Why not just combine the two cases? Is there any harm in allowing the "md5" case to check for a "final_digest" annotation?

byronlin13 replied:
Well we are updating different arrays. For MD5 we don't have md5FinalDigest list

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,536 @@
+	hr, err := NewHashReader(reader, provider)
+	require.NoError(t, err, "Something went wrong executing NewHashReader()")
+	TraverseReader(hr)
+	hr.Sum(nil)
```
Are we doing anything with this sum result or is this just for the purpose of updating the hashlog?

byronlin13 replied:
Just for the purpose of updating the hashlog

justing-bq replied:
Lets handle the hr.Sum(nil) error in assert.NoError()

byronlin13 replied:
Updated

R-maan commented on identity_hasher.go:
```diff
@@ -19,26 +19,31 @@ type identityHasher struct {
+func (ih *identityHasher) Sum(bytes []byte) []byte {
 	// We ignore the error here because we know this particular Write() implementation does not error
-	_, _ = identityHasher.Write(bytes)
+	_, _ = ih.Write(bytes)
```
I believe you can just write ```ih.Write(bytes)``` and leave out ```_, _```

byronlin13 replied:
Goland complains about `unhandled error` when I try that

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,220 @@
+
+		assert.NoError(t, reader.StepIn(), "Something went wrong executing reader.StepIn()")
+
+		reader.Next() // Read the initial Ion value.
```
You should put this in an assert.True()

byronlin13 replied:
Addressed.

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,220 @@
+		}
+
+		// Iterate through expected/ digest bytes.
+		reader.Next()
```
And here.

byronlin13 replied:
Addressed.

justing-bq commented on ion_hash_test.go:
```diff
@@ -0,0 +1,220 @@
+		Traverse(t, reader, provider.getInstance())
+
+		if len(parameters[i].expectedHashLog.identityUpdateList) > 0 {
+			assert.Equal(t, provider.getUpdateHashLog(), parameters[i].expectedHashLog.identityUpdateList, parameters[i].hasherName+" failed")
```
The expected value should go before the actual value. Switch these around.

byronlin13 replied:
Updated
